File: src/main/java/pramp/FlattenDictionary.java
package main.java.pramp;

import java.util.HashMap;
import java.util.Map;

public class FlattenDictionary {
    static HashMap<String, String> flattenDictionary(HashMap<String, Object> dict) {
        // your code goes here
        HashMap<String, String> result = new HashMap<>();
        helperFn("", dict, result);
        return result;
    }

    static void helperFn(String initialKey, HashMap<String, Object> dict, HashMap<String, String> result) {
        for (Map.Entry<String, Object> entry : dict.entrySet()) {
            String key = entry.getKey();
            Object value = entry.getValue();

            String newKey = initialKey.isEmpty() ? key : initialKey + "." + key;
            // If the value type is HashMap, call the helper fn recursively on it
            if (value instanceof HashMap) {
                helperFn(newKey, (HashMap<String, Object>) value, result);
            } else {
                // Base case: value is a String, concatenate it to the result
                result.put(newKey, (String) value);
            }
        }
    }
}

File: src/main/java/pramp/FlattenDictionary.kt
package main.java.pramp

fun flattenDictionary(dict: Map<String, Any>): Map<String, String> {
    val result = mutableMapOf<String, String>()
    flattenDictionaryHelper("", dict, result)
    return result
}

fun flattenDictionaryHelper(initialKey: String, dict: Map<String, Any>, result: MutableMap<String, String>) {
    for ((key, value) in dict) {
        val newKey = when {
            initialKey.isEmpty() -> key
            key.isEmpty() -> initialKey
            else -> "$initialKey.$key"
        }

        if (value is Map<*, *>) {
            // Recursive call with casting to the correct type
            @Suppress("UNCHECKED_CAST")
            (flattenDictionaryHelper(newKey, value as Map<String, Any>, result))
        } else {
            // Base case: value is a String
            result[newKey] = value.toString()
        }
    }
}

fun main() {
    val dict = mapOf(
        "Key1" to "1",
        "Key2" to mapOf(
            "a" to "2",
            "b" to "3",
            "c" to mapOf(
                "d" to "3",
                "e" to mapOf(
                    "" to "1"
                )
            )
        )
    )

    println(flattenDictionary(dict))
}